File: ports/1.21.1-full/src/main/java/mods/eln/server/ElnWorldStorage.java
package mods.eln.server;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.level.Level;
import net.minecraft.world.storage.MapStorage;
import net.minecraft.world.level.saveddata.SavedData;

public class ElnWorldStorage extends SavedData {

    private int dim;

    final static String key = "eln.worldStorage";

    public ElnWorldStorage(String str) {
        super(str);
    }

    public static ElnWorldStorage forWorld(Level world) {
        // Retrieves the MyWorldData instance for the given world, creating it if necessary
        MapStorage storage = world.getPerWorldStorage();
        int dim = world.provider.getDimension();
        ElnWorldStorage result = (ElnWorldStorage) storage.getOrLoadData(ElnWorldStorage.class, key + dim);
        if (result == null) {
            result = (ElnWorldStorage) storage.getOrLoadData(ElnWorldStorage.class, key + dim + "back");
        }
        if (result == null) {
            result = new ElnWorldStorage(key + dim);
            result.dim = dim;
            storage.setData(key + dim, result);
        }
        return result;
    }

    @Override
    public void readFromNBT(CompoundTag nbt) {
        dim = nbt.getInteger("dim");
        ServerEventListener.readFromEaWorldNBT(nbt);
    }

    @Override
    public CompoundTag writeToNBT(CompoundTag nbt) {
        nbt.setInteger("dim", dim);
        ServerEventListener.writeToEaWorldNBT(nbt, dim);
        return nbt;
    }

    @Override
    public boolean isDirty() {
        return true;
    }
}

File: src/main/kotlin/mods/eln/server/ElnWorldStorage.kt
package mods.eln.server

import mods.eln.misc.DimensionIds
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.Level
import net.minecraft.world.level.saveddata.SavedData

/**
 * The legacy per-dimension world storage of the node graph (the fallback when the mod's own
 * `electricalAgeWorld<dim>.dat` files cannot be read). 1.21: a SavedData per server level.
 */
class ElnWorldStorage private constructor(private val dim: Int) : SavedData() {
    override fun save(nbt: CompoundTag, registries: HolderLookup.Provider): CompoundTag {
        nbt.putInt("dim", dim)
        ServerEventListener.writeToEaWorldNBT(nbt, dim)
        return nbt
    }

    override fun isDirty(): Boolean {
        return true
    }

    companion object {
        const val key = "eln.worldStorage"

        private fun factory(dim: Int) = Factory({ ElnWorldStorage(dim) }, { nbt, _ ->
            val d = if (nbt.contains("dim")) nbt.getInt("dim") else dim
            ServerEventListener.readFromEaWorldNBT(nbt, d)
            ElnWorldStorage(d)
        }, null)

        @JvmStatic
        fun forWorld(world: Level): ElnWorldStorage? {
            val level = world as? ServerLevel ?: return null
            val dim = DimensionIds.id(level)
            return level.dataStorage.computeIfAbsent(factory(dim), key + dim)
        }
    }
}
